File: problem-solving/binary-search.js

const binarySearch = function (nums, target) {
    let left = 0;
    let right = nums.length - 1;

    while (left <= right) {
        const middle = left + Math.floor((right - left) / 2);

        if (target < nums[middle]) {
            right = middle - 1;

        } else if (target > nums[middle]) {
            left = middle + 1;

        } else {
            return middle;
        }
    }

    return -1;
}

function binarySearchRecursive(ar, target, left = 0, right = ar.length - 1) {
    const middle = Math.ceil((right + left) / 2);

    if (ar[middle] == target) {
        return middle
    }

    if (ar[middle] > target) {
        return binarySearch(ar, target, left, middle - 1);
    }

    return binarySearch(ar, target, middle + 1, right);
}

const arr = [1, 2, 3, 4, 5, 6, 7];
console.log('=====================');
console.log(binarySearch(arr, 3));
console.log(binarySearchRecursive(arr, 3));
console.log('=====================');

File: problem-solving/binary-search.py

def binary_search(nums, target: int, start=0, end=None) -> int:
    if end is None:
        end = len(nums) - 1

    middle_index = int((start + end) // 2)

    if target == nums[middle_index]:
        return middle_index

    if target < nums[middle_index]:
        return binary_search(nums, target, start, middle_index)

    else:
        return binary_search(nums, target, middle_index, end)


print(binary_search([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 7))
